### ISSUE TITLE
Job Controller

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#1624]
We need something like ReplicationController that runs RestartOnFailure and RestartNever pods "to completion", collects results, etc

**Comment by TamerTas** [comment]
@thockin @smarterclayton @bgrant0607 After a long journey through the issues about job management in kubernetes, I have submitted a GSOC proposal for this topic. Your feedback would be most appreciated.

**Comment by bgrant0607** [comment]
We'll take a look @TamerTas. Thanks!

**Comment by bgrant0607** [comment]
To document later:

Some batch workloads are data processing/analysis workloads of independent utility.

Others support serving workloads, such as:
* data cleanup / GC
* serving data generation / aggregation / indexing / import
* data snapshots / copies / backups / export
* logs processing / billing / audit / report generation
* maintenance (e.g., schema changes / data conversion)
* integrity checking / validation
* defense analysis (spam, abuse, dos, etc.)
* online/offline feedback / adaptation / machine learning
* continuous/periodic build/push

**Comment by TamerTas** [comment]
@bgrant0607 I took your considerations regarding my proposal and tried to answer them could you take another look?

**Comment by bgrant0607** [comment]
Minimal version:
* Use ReplicationController as starting point
* RestartOnFailure only
* user specifies desired number completed successfully
* persistently track successful completions in annotation (at least for now), which is reflected in status

cc @davidopp

**Comment by derekwaynecarr** [comment]
Do you view a scheduled time to run as something above a JobController?  For example, if I want to run scheduled jobs every day/month/year, is that something different than this JobController and another layer?  We have been discussing this on our side and wanted your take based on your experience.

**Comment by bgrant0607** [comment]
Yes, I view scheduled/periodic jobs as a separate controller -- #2156.  I might consider putting deadline scheduling into this controller.

**Comment by derekwaynecarr** [comment]
I am interested in making a proposal and implementation soon that would let me set a max runtime duration (In seconds) for a container in a RunOnce pod that would be enforced by the Kubelet.  If pod was not done it would be killed and marked failed.  I want to make sure that that is distinct from what you referred to as deadline scheduling.  We need this concept to time box certain activities in near term.  Think we talked about this at last meetup, wanted to make sure your thinking had not changed.

**Comment by derekwaynecarr** [comment]
Sent last note too quickly but soon meant < 2 weeks out.

**Comment by bgrant0607** [comment]
@derekwaynecarr Yes, that's fine. It would be a field on the pod.

**Comment by davidopp** [comment]
@bprashanth This is the issue I was talking to you about earlier. See @bgrant0607 's earlier comment here: https://github.com/GoogleCloudPlatform/kubernetes/issues/1624#issuecomment-95357364

To flesh out what he said just a little more (based on discussion with him earlier), the idea is that the JobController runs some number of pods to completion. Kubelet is responsible for restarting failed pods locally, and if the machine dies the controller is responsible for restarting the failed pod on another machine. For now, we want the JobControllerSpec to specify the number of instances to run to completion (call it N); the controller is free to decide how many run in parallel at any one time, but the most obvious thing to do is to start up N replicas initially and create a new pod when an existing pod fails (but not when it succeeds). So:
* clone ReplicationController, rename it JobController, and then edit
* validation should ensure that PodSpec.RestartPolicy is RestartPolicyOnFailure. Other policies don't make sense here.
* instead of Replicas field like in ReplicationController, you want something like DesiredNumCompleted (self-explanatory)
* JobControllerSpec should have an annotation that tracks number of pods that have successfully completed, and this should be reflected in the JobControllerStatus. JobController should update JobControllerStatus and the annotation together, so that the user's view of # completed always matches the controller's view. 
* the documentation should warn users that it is possible, albeit unlikely, that more than N instances will run to completion, because it's always possible that a Kubelet will crash after a container exits with success but before Kubelet can inform the master.

**Comment by bgrant0607** [comment]
And we should call it Job, not JobController.

**Comment by davidopp** [comment]
Definitely!

**Comment by bprashanth** [comment]
Plan to start on this right after my current rc manager rewrite gets in (hopefully EOW). I'll start by adding a new Job resource with validation strategy, associated etcd storage etc, and follow up with a controller half that watches jobs and pods, and takes action they succeed/fail.

**Comment by erictune** [comment]
I've looked at #7380.  I don't understand how one would use the Job resource to run an actual multi-pod computation?  

Any multi-pod computation I can think of has to consist of multiple work units.  The computation is done when all the work units are done.   

There are two general ways I can see to assign work units to workers (pods): via a varying pod spec, or via an external mechanism (a work queue).

Neither one works well with the proposed Job resource.

In the work-queue model, when a pod starts, it talks to something that tells it what work unit to work on.   If it a pod completes its work unit, there is no reason it can't get another work unit.   This can help minimize overall job runtime when work units are unevenly sized.  And it amortizes pod startup costs.  For this model, you want a fixed time or cpu/memory or pod-count budget, and let the controller decide how many pods to run, until the work units are all done.  Pod completion count doesn't matter.

In the varying-pod-spec model, there is a 1:1 mapping between pods and work units.  Each pod has a different spec (different flags or different env vars, or something).  I don't see how you can do this with the proposed job resource.

**Comment by bprashanth** [comment]
> I've looked at #7380. I don't understand how one would use the Job resource to run an actual multi-pod computation?

The proposed resource will only allow a single pod spec to run to completion n times. It essentially implements:

> For now, we want the JobControllerSpec to specify the number of instances to run to completion (call it N); the controller is free to decide how many run in parallel at any one time, but the most obvious thing to do is to start up N replicas initially and create a new pod when an existing pod fails (but not when it succeeds).

Long lived pods that pick off of a work queue sounds like an rc use case? I can think of several other use cases where I'd want something to run once with infinite retries till completion (database purge, sending a confirmation email post sign up etc). One could also extend the db cleanup example to spin off 5 jobs that purge different tables.

**Comment by erictune** [comment]
If there is no work queue and the pods have a single spec, then won't they be doing redundant computations?

**Comment by bprashanth** [comment]
You can still have a workqueue, just that the cost of keeping these replicas running is too high given the size of the queue. For example you could have a job with completions=1 create a workqueue with 5 items, and create another job with completions=5. If this operation is going to happen once a month it wouldn't justify keeping the workers around?

I agree that the single spec, no work queue, N>1 completions case doesn't sound very useful, but giving a job with N>1 completions different pod templates can also be achieved with N jobs that have completions=1.

**Comment by erictune** [comment]
Can you give an example of a use case where completions needs to be greater
than 1?

**Comment by bprashanth** [comment]
The use case above? (i.e a workqueue that is created on demand every month by another job)

**Comment by erictune** [comment]
So, the Job resource can only be used when # work units == # completed pods.  

I see that you can make that work.  But I don't think it is a good fit for people's mental models of parallel computation.

**Comment by davidopp** [comment]
I think @erictune is making a good point. I don't think we really spec'd this correctly. As he says, pods need some way of knowing which data they should compute on. The way to do this is to give each of the pods that is started up initially a unique identifier that remains constant across restarts (on the same machine or on another machine). This can be an index or even a random value from a large keyspace (like the consistent hashing technique used to partition data in peer-to-peer distributed hash tables). I think this is where the "everything is pure cattle" model breaks down a little, but it seems unavoidable. Any time you have a computation where you partition the data, workers are not identical. I think this is true whether you are using a "worker pool" model or a more traditional partitioned data model.

**Comment by smarterclayton** [comment]
Which is similar, but not identical to, nominal services.  They should at least feel similar even if they are distinct entities.

**Comment by alex-mohr** [comment]
.. if this isn't 1.0 blocker and isn't well-enough specified, perhaps drop this in favor of another issue and come back to it later?

**Comment by davidopp** [comment]
@bprashanth and I had a fairly fruitful discussion about this today. I am going to try to summarize.

First, we both agreed that discussion of this issue is suffering from our not having started with very concrete use cases. To help our discussion, we decided there are fundamentally two models of batch computation, distinguished by the answer to this question: do workers communicate/coordinate after startup.
* Example where the answer is "no": A large read-only dataset is stored in a cluster filesystem (say, as rows of a table). N workers are created and are assigned at startup time (e.g. in the command line args) a unique index X between 0 and N-1; a worker processes row # X, X+N, X+2N, X+3N, X+4N, ...
* Example where the answer is "yes": A large read-only dataset is stored in a cluster filesystem. But this time, there is also a way for workers to let each other know which rows they have processed/are processing. For example, they may have read-write access to a second file where they record each row number as they work on it. This is representative of any work queue system where workers can communicate (with one another or with a coordination point like a queue service).

The job controller that @bprashanth implemented appears to work for the second type of batch job. But it doesn't work for the first type. The first type requires giving each worker a unique index and the total number of workers (there are many ways to do this -- multiple pod templates, a single pod template with argument substitution/per-instance overrides, etc.)

So the questions we need to answer are
* do we want to do any kind of job controller for 1.0
* if yes, is it sufficient to say we just support the second type for now, and argue about how to support the first type later, or do we want to cover both cases at the same time
* if yes, should we treat this as an "experimental" API (an idea @brendandburns has proposed) so we can iterate on it before locking into it
* is this proposed dichotomy complete, i.e. are there other meta-use-cases we need to consider

We also noted some relationship between this issue and #6393. The proper building block for a job of the first type is the kind of pod @alex-mohr proposed--something that has a unique identity and keeps working on its partition of the input data even if it gets rescheduled to a new machine. The Borg analogy (for those familiar) is that a Borg Job is built on top of Tasks and each Task is built on top of Attempts. Building a Kubernetes JobController directly on top of today's Pods is like building a Borg Job on top Attempts -- you can do it, but the controller that implements the job is going to end up internally implementing an abstraction like a Task (or Alex's Pod). And the user definitely wants an Alex-Pod/Borg Task abstraction for reasoning about/monitoring/etc. their job. (Again, everything I said in this paragraph applies to the first type of job.)

**Comment by derekwaynecarr** [comment]
Thanks for this write up. 

It was helpful.  

For Job like tasks we have discussed, we are more interested in scheduled jobs (which would happen above what is proposed here for things like cron) and more single pod worker style jobs that are mainly just run once pods that need to run to success (think a build of source code).

**Comment by davidopp** [comment]
@derekwaynecarr Yeah I think cron is an orthogonal dimension like you said; it's what creates the Job, and doesn't really constrain the semantics of the Job itself one way or the other. I don't know the details of your build-source-code use case; if each pod can pull the description of what work it needs to do from some kind of queue service, then it could be viewed as "type 2," but the work assignment is statically determined, then it seems more like "type 1."

**Comment by smarterclayton** [comment]
In our case the job is fully parameterized by the creator of the job (it's maybe type 0, needs no input beyond its arguments).

**Comment by davidopp** [comment]
What difference, if any, is there between each replica (pod) of one of your jobs? I think it would be "type 1" if each replica has different arguments/parameters.

**Comment by smarterclayton** [comment]
No difference between them.  Job is done once (just need one to run to completion).

The build is "grab a source repo, built the commit we were parameterized with, push to registry when done".  Each build is its own distinct job.

**Comment by davidopp** [comment]
I see. So you have the "run to completion, not forever/daemon" property but not the "multiple replicas" property. So yeah, I think @gmarek's approach would work for your use case (with num_completions == 1). It sounds like @alex-mohr's proposal in #6393 would also satisfy your use case, without requiring any kind of explicit controller or new abstraction (not saying we should necessarily build that, just observing).

**Comment by smarterclayton** [comment]
Right - we implemented "job" (run pod multiple times) on top of pod.  Every pod in the job is the same, so not quite type 1 or type 2.  We would expect to use "Job" however implemented to prevent us from having implement it ourselves.  Our use case is probably the simplest possible Job so it's not really relevant to this discussion other than as a minimum bar.

----- Original Message -----

**Comment by wojtek-t** [comment]
/sub
cc @fgrzadkowski

**Comment by erictune** [comment]
@smarterclayton if every pod is the same in your scenario, what makes them do different things?

**Comment by smarterclayton** [comment]
For an individual job, nothing.  The goal is to get at least one success (by running the pods in succession until one is identified as reaching completion).  So two pods that run to completion under a single job is acceptable, but zero pods is not.

**Comment by nwendling** [comment]
We have this task in Openshift Origin plateform Experience (https://trello.com/c/7xCibp7B/374-8-as-an-openshift-operator-i-need-create-run-once-pods-for-recurrent-execution-future) about been able to execute pods with a finite lifetime, basically in our case to be able to schedule batch executions.

Having exchanged a few mails with @derekwaynecarr, our initial idea was to introduce some kind of time based job scheduler object directly in Kubernetes, but it looks like this is something that should better be implemented in an upper layer. However for a cron-like applications to be able to rely on kubernetes for tasks execution, having this new job concept could definitely help. 

To me, this job should look like an construction being able to aggregate the execution of one or several pod templates, each started with a configured number of instances, to track the outcome of each started pod and collect its exit status, and be able to advertise a final completion status once all pods finished, offering to external application relying on Kubernetes a way to watch this outcome. 

Currently there is nothing similar in kubernetes. I guess an external application could still possibly do something like POSTing a pod template definition, watching pod status changes until completion and then perform the necessary cleanup (or forget about that step ...). 

This new job object could therefore offer some basic fonctionalities to avoid having them being re-implemented by each client application:

- Start pods based on one or several templates, each started with a configured number of instances.
- Report a unique completion status after completion of all pods started for this job execution.
- Provide restart on failure (up to a maximal count)/maximal execution time features, defined at pod/job level.
- Automatically infect all started pods with environnment variables providing the name of the job, the number of started pods for a given type, the pod instance index from the whole list.

The job object would start its execution as soon as created in the system (we could maybe think about adding some delay startup).
Once completed, a default TTL could be set to its associated ETCD key for automatic cleanup. 
We could think about implementing something similar to the waitpid routine, a WATCH entry for jobs that would reply to client the final state of this job object only upon completion and immediately remove this job object from the registry ?

I don't know if such a design could make sense ?

**Comment by davidopp** [comment]
Yes, what you described is basically what we had in mind (though we hadn't thought much about aggregating a single job-level completion status from the pod-level termination statuses).

**Comment by bgrant0607** [comment]
My goal was to add the minimal useful Job abstraction. If just a single instance would be less controversial, I'd be happy to start with that.

I agree that we need an experimental API prefix. This comes up over and over again.

For the controller implemented by @bprashanth, I imagined that any coordination necessary would take place in the application, at least for the initial implementation.

As with nominal services, I disagree that a primitive like that proposed in #6393 is necessary or even helpful in building a multi-instance abstraction, and I strongly recommend that we don't rathole on that here.

**Comment by davidopp** [comment]
> For the controller implemented by @bprashanth, I imagined that any coordination necessary would 
> take place in the application, at least for the initial implementation.

As I alluded to above, I don't think a job controller that only supports this model (type 2 from my earlier comment) is useful. I think even the initial version needs to support pets (type 1) in the manner @nwendling and I have described in earlier comments.

> I disagree that a primitive like that proposed in #6393 is necessary or even helpful in building a multi-
> instance abstraction

It seems very clear to me that users with type 1 batch jobs will want to observe status (and more generally, reason about their job) in terms of an abstraction like the one @alex-mohr proposed in #6393 (a worker that runs to completion and is restarted by the system on the same or different node when it fails, while keeping its identity across these restarts). But we don't have to build the feature/object @alex-mohr described.

**Comment by bgrant0607** [comment]
A single-instance job would be a subset of any functionality we'd add later, and would be useful on its own. I suggest starting with that.

For more general types of jobs, let's use mapreduce as an example:
http://research.google.com/archive/mapreduce.html

It has a "master", which one could think of as a controller. The master assigns work, supervises workers, starts duplicate workers for "stragglers", and more. It multiplexes work onto the workers. The master doesn't need dense, stable task indices. It just needs a queue of incomplete work and a scoreboard of current assignments to identifiable workers.  Only the master can provide current work completion status information.

Many similar systems have been developed over the years, but the majority have a similar master/worker design, or at least a queue/scoreboard manager. A standard lease-based shared task queue that requires consumers to notify the queue upon task completion would also be sufficient for many use cases.

Indices have value mainly where they can fully replace dynamic work assignment, for performance or complexity reasons, and where consistent hashing isn't viable.

For >1 pod, I agree that a concrete use case would be useful. [Dataflow](https://github.com/GoogleCloudPlatform/DataflowJavaSDK) might be a worth a look, or a use case brought by an actual user.

**Comment by bgrant0607** [comment]
With respect to recurring workloads, I'll again redirect to #2156.

**Comment by davidopp** [comment]
I guess I'm OK with initially just supporting "type 2" (or the even simpler version of "type 2" where num_completions=1), especially if it would allow us to do something for batch for 1.0, but I think we should make sure we have a plan for supporting "type 1" soon after that, and try to avoid designing the "type 2" API in a way that we have to change to later support "type 1" (though I agree using the experimental API prefix would make it less bad to change later).

I think we disagree about the relative importance of type 2 vs. type 1, but that probably doesn't matter (as long as we eventually implement support for both).

**Comment by bgrant0607** [comment]
Instead of types 1 and 2, let's call them static and dynamic work assignment.

I don't disagree that static work assignment could be useful. I view it as similar to nominal services (#260), but where the identity/role would be assigned based on the uncompleted task assigned rather than which currently running instance it was (keeping number of completed tasks desired orthogonal to max/current number of pods performing tasks).

I don't think that it's a requirement for a minimal viable API, however.

**Comment by bgrant0607** [comment]
cc @erictune

**Comment by gmarek** [comment]
cc @mwielgus

**Comment by smarterclayton** [comment]
@soltysh please link your ongoing job proposal here so Kube folks can get a chance to look.  The proposal will be coming soon here while Maciej prototypes

**Comment by soltysh** [comment]
It's already linked here, but here you go https://github.com/openshift/origin/issues/3693. It covers both the job controller part and the cron scheduler.

**Comment by bgrant0607** [comment]
Could we discuss this at an upcoming community hangout?

**Comment by smarterclayton** [comment]
Absolutely

On Jul 21, 2015, at 3:06 PM, Brian Grant <notifications@github.com> wrote:

Could we discuss this at an upcoming community hangout?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/1624#issuecomment-123445893>
.

**Comment by bprashanth** [comment]
I'm not working on the job controller, a better owner might be Mike or @soltysh

**Comment by davidopp** [comment]
@erictune has been shepherding this, so assigning to him.

**Comment by bgrant0607** [comment]
@aronchick Continuing from https://github.com/kubernetes/kubernetes/issues/14186#issuecomment-144841868

Your primary concern is to clarify that this is a batch job (e.g., [LSF](https://www-01.ibm.com/support/knowledgecenter/SSETD4_9.1.3/lsf_users_guide/job_submit.dita), [Load Leveler](https://docs.loni.org/wiki/Useful_LoadLeveler_Commands#Submit_a_Batch_Job), [Printer Job](http://docs.oracle.com/javase/7/docs/api/java/awt/print/PrinterJob.html)) as opposed to an indefinitely running "Job" (e.g., [Borg](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf), [Aurora](http://aurora.apache.org/documentation/latest/configuration-reference/#job-schema), [Nomad](https://www.hashicorp.com/blog/nomad.html))?

**Comment by soltysh** [comment]
And we do distinguish those, but as:
- job - run-once pods
- replicationcontroller (to be renamed someday #3024) - run indefinitely pods

That's the terminology we have in k8s. @aronchick what's your opinion on that?

**Comment by erictune** [comment]
We now have this as a beta feature, in head and to appear in 1.1.  Closing.

**Comment by soltysh** [comment]
@erictune are you going to close it?
